Fix out of bound read in CaosLexer

mtlynch opened the pull request:
I noticed that the CaosLexer has some bugs where it reads beyond the null termination character for malformed input. This can cause a read into unallocated memory, which can potentially cause the code to crash.

There's a security concern here, as malicious input can cause the lexer to read uninitialized memory (similar to [Heartbleed](https://en.wikipedia.org/wiki/Heartbleed)), though it's only a single byte, and it's hard to imagine an attacker weaponizing this in the context of openc2e.

This change does a few things:

1. Adds a unit test to exercise the previously vulnerable code path. You can't see the failure since the fix is also in this PR, but you can see what happens without the fix [in this build on my copy of the repo](https://github.com/mtlynch/openc2e/actions/runs/11849613968/job/33023157739)
1. Fixes the vulnerability by gracefully handling the null terminator following a double quote character

mtlynch commented:
There are several similar bugs in the CaosLexer.

Here's [a preview](https://github.com/openc2e/openc2e/compare/main...mtlynch:openc2e:fix-heap?expand=1) of all the fixes.

Is it helpful if I pull in the rest of the fixes into this PR? Or should I do one PR per fix to make it obvious which unit test exercises which fix?

mtlynch commented:
@ligfx - Anything I can do to make this PR easier to review?

I know you're not obligated to take my code, but I'd love to contribute, so I'm happy to adjust in a way that reduces maintenance burden for you.

ligfx commented:
@mtlynch thanks for the ping.

- Note that `-fsanitize=address` is already added to `CMAKE_CXX_FLAGS` in `CMakeLists.txt` above where you added it. That logic might need to be tweaked if it doesn't work on your environment.
- No need to create Debug builds as ASAN is added unconditionally right now.
- Can you clean up/rebase the commits so each is one logical change?

mtlynch commented:
@ligfx Thanks for reviewing!


>Note that -fsanitize=address is already added to CMAKE_CXX_FLAGS in CMakeLists.txt above where you added it. That logic might need to be tweaked if it doesn't work on your environment.
 >No need to create Debug builds as ASAN is added unconditionally right now.
 
Oh, I missed that. I've fixed it, so that simplifies things a lot.

>Can you clean up/rebase the commits so each is one logical change?

Sure, I've squashed it down to a single commit.

ligfx commented:
Thanks @mtlynch , merged.

Made a couple tweaks you can see in 981b9cae6fdc365b72cbdd23c8b2058290496a38:
- apply `clang-format`
- Don't need to `goto eoi;` after `push_value`, it goes back to the top of the function automatically
- Turns out fmt already can do a debug representation for strings with `{:?}`, so I just used that

mtlynch commented:
Awesome, thanks!
